cdm/core/src/main/java/ucar/nc2/filter/ScaleOffset.java

```diff
@@ -166,26 +249,14 @@ private static ByteOrder parseByteOrder(String dtype, ByteOrder defaultOrder) {
     return defaultOrder;
   }
 
-  private DataType nextLarger(DataType dataType) {
-    switch (dataType) {
-      case BYTE:
-        return SHORT;
-      case UBYTE:
-        return USHORT;
-      case SHORT:
-        return INT;
-      case USHORT:
-        return UINT;
-      case INT:
-        return LONG;
-      case UINT:
-        return ULONG;
-      case LONG:
-      case ULONG:
-        return DOUBLE;
-      default:
-        return dataType;
-    }
+  public double applyScaleOffset(Number value) {
+    double convertedValue = value.doubleValue();
+    return Math.round((convertedValue - offset) * scale);
```

Comment:
Should this be
```suggestion
    return (convertedValue - offset) * scale;
```
since the `scale` is now a `double`?

Reply:
ooops

Reply:
I see now that removing the round will give incorrect results for non-floating point types, as they will later get cast, so the floor will be taken. However, with the round you get incorrect results for floating point scale/offsets.

Maybe a solution could be
```suggestion
    if (astype == DOUBLE || astype == FLOAT) {
      return (convertedValue - offset) * scale;
    }
    return Math.round((convertedValue - offset) * scale);
```